### PR TITLE
feat(a11y-query): extend aria handler with waitFor

### LINKED ISSUE #6307: Proposal: Query by accessible name
## Motivation
Currently Puppeteer supports querying the DOM using CSS selectors. While this provides a versatile way to select DOM elements by for example id, tag, or class-attributes, these properties are often subject to change or even generated at build-time.
Alternatively, we could support querying the accessibility tree for accessible names.
This could provide the following benefits:
- Make selectors in test scripts more resilient to source code changes.
- Make test scripts more readable since the accessible names should carry semantic context.
- Motivate good practices for assigning accessibility properties to elements.

## Current state of affairs
Querying by accessible names is currently possible by implementing a custom query handler to traverse the DOM and check accessible names along the way. This is what [puppeteer/recorder](https://github.com/puppeteer/recorder) currently does. Since there is no direct way to get the computed accessible name and role from the browser-side, one has to compute those on demand which is not a trivial task.
One can instead access and traverse the accessibility tree through Puppeteer where the computed accessible names and roles are available. The catch then is that Puppeteer only gives access to a static snapshot of the accessibility tree. 
To support interaction, we need a mapping to live DOM nodes.

## Proposal
Extend Puppeteer to support querying the accessibility tree with support for interaction.
The steps for achieving this could be:
- [x] Implement CDP-level support for querying the accessibility tree.
Landed in https://github.com/chromium/chromium/commit/d557b1cd92b3c47dae01aaaddd7c6be0ceb6cb94
- [x] Restructure such that query handlers can query on the Puppeteer-level (as opposed to querying in the page context) 
Done in #6437.
- [x] Implement atlernative query handler for a11y-based querying
Done in #6459  and #6472.

### PR DESCRIPTION
**Pull request by johanbay**
This PR adds `waitFor` to the built-in aria handler (#6307).
This is part 3 of https://github.com/puppeteer/puppeteer/pull/6453.

### CODE REVIEW

File: src/common/AriaQueryHandler.ts
Comment (johanbay) [inline src/common/AriaQueryHandler.ts]: Is this a reasonable way to implement a semaphore-like waiting logic between asynchrounous operations?

File: src/common/AriaQueryHandler.ts
Comment (mathiasbynens) [inline src/common/AriaQueryHandler.ts]: This is complicated and a little magical to me, so apologies if I misunderstood, but: would something like this work here?

```js
hasBinding = true;
```

instead of `settingUpBinding` and `done`

File: src/common/DOMWorld.ts
Comment (mathiasbynens) [inline src/common/DOMWorld.ts]: `window` -> `globalThis` like elsewhere in the PR? That way we support worker contexts as well. Even if those don't need to be supported in this code path, it's more consistent with other code in this PR

**Review by mathiasbynens — COMMENTED** [review]
Not done reviewing, but some initial comments before I sign off

File: src/common/AriaQueryHandler.ts
Comment (johanbay) [reply src/common/AriaQueryHandler.ts]: I'm not sure how that would work. Suppose we do the following
```addHandlerToWorld(...); await addHandlerToWorld(...);```
The first call starts adding the binding, but at some point it is blocked while waiting for the CDP response. Now, the second call can either decide that the binding has already been added (which it hasn't yet) or itself start adding the binding as well.
The code here tries to fix this situation by having the first call set up a promise in `settingUpBinding` which it resolves when it is done adding the binding by calling `done()`. The second call sees that the other call is setting up the binding and blocks until the `settingUpBinding` promise is resolved.

Is there a simpler (or more canonical) way to achieve this?

Comment (mathiasbynens) [reply src/common/AriaQueryHandler.ts]: I see — we don't just need to know whether or not the binding has been set up, we want to know whether it's already _in the process of being set up_.

I know this is captured in your code comment above but somehow I didn't understand the full intent behind this code until just now — thanks for the explanation!

File: src/common/DOMWorld.ts
Comment (mathiasbynens) [inline src/common/DOMWorld.ts]: Let’s make sure this produces a correct message for arbitrary string values for `name`s (e.g. if `name` includes a `'`):

```suggestion
        `Failed to add page binding with name ${name}: window[${JSON.stringify(name)}] already exists!`
```

Similar to the other comment, we could use `globalThis` instead of `window` — WDYT?

File: src/common/AriaQueryHandler.ts
Comment (OrKoN) [inline src/common/AriaQueryHandler.ts]: Can you do `settingUpBinding  = domWorld.addBinding(...)` and then `await settingUpBinding`? I think it might help avoid having the extra promise?

File: src/common/DOMWorld.ts
Comment (mathiasbynens) [inline src/common/DOMWorld.ts]: nit:

```suggestion
          'Execution context is not available in detached frame'
```

File: src/common/AriaQueryHandler.ts
Comment (OrKoN) [reply src/common/AriaQueryHandler.ts]: So bindings are set up per world? What would happen if we do `addHandlerToWorld(world1); addHandlerToWorld(world2)`? Is it possible to add bindings to two different worlds concurrently?

File: src/common/helper.ts
Comment (mathiasbynens) [inline src/common/helper.ts]: I don't like relying on function hoisting — WDYT about moving the `return` to be the last line in this function body?

File: test/ariaqueryhandler.spec.ts
Comment (mathiasbynens) [inline test/ariaqueryhandler.spec.ts]: What does it look like currently? Can you make a recommendation?

File: src/common/AriaQueryHandler.ts
Comment (johanbay) [reply src/common/AriaQueryHandler.ts]: Ah, good point! In practice, bindings are set up per execution context, but the world is responsible for re-binding when the execution context changes. It's fine to add bindings in two different worlds concurrently (it should happen when doing `waitForSelector` in two different frames), so maybe this code actually synchronizes a bit too much. I'll try pushing it into the worlds instead.

File: src/common/DOMWorld.ts
Comment (johanbay) [reply src/common/DOMWorld.ts]: Done. Also updated the actual error.

File: src/common/DOMWorld.ts
Comment (johanbay) [reply src/common/DOMWorld.ts]: Good idea, done.

File: src/common/helper.ts
Comment (johanbay) [reply src/common/helper.ts]: Done.

File: src/common/AriaQueryHandler.ts
Comment (johanbay) [reply src/common/AriaQueryHandler.ts]: Done, but moved into DOMWorld instead.

File: test/ariaqueryhandler.spec.ts
Comment (johanbay) [reply test/ariaqueryhandler.spec.ts]: This is the current stack trace:
```
"TimeoutError: waiting for selector `zombo` failed: timeout 10ms exceeded
    at new WaitTask (.../puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:494:34)
    at DOMWorld.waitForSelectorInPage (.../puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:427:26)
    at Object.waitFor (.../puppeteer/lib/cjs/puppeteer/common/AriaQueryHandler.js:55:21)
    at process._tickCallback (internal/process/next_tick.js:68:7)"
```
It would be nice if `ariaqueryhandler.spec.ts` was part of the trace since that is the actual call site for `waitForSelector`.
The fact that the selector is `zombo` instead of `aria/zombo` is maybe a bit unfortunate as well.

File: src/common/AriaQueryHandler.ts
Comment (johanbay) [reply src/common/AriaQueryHandler.ts]: Moved to `DOMWorld.ts` to allow asyncronously setting up bindings in multiple worlds.

File: src/common/DOMWorld.ts
Comment (johanbay) [reply src/common/DOMWorld.ts]: Done.

File: test/ariaqueryhandler.spec.ts
Comment (johanbay) [reply test/ariaqueryhandler.spec.ts]: It seems like we can't do much better than this since waitFor is async.

File: src/common/DOMWorld.ts
Comment (OrKoN) [inline src/common/DOMWorld.ts:491]: Would it set up the binding for the same world again?

Comment (johanbay) [reply src/common/DOMWorld.ts:491]: No it shouldn't. The first check `this._ctxBindings.has(name)` should be true if the binding was added by the current operation, so it will return immediately.
The idea behind the recursive call is that if multiple tasks are blocked on `await this._settingUpBinding;`, then one of them will get to do the recursive call and assign a new promise to `_settingUpBinding`. When the other tasks gets to do the recursive call they are therefore blocked here again.

**Review by OrKoN — APPROVED** [review]

**Review by mathiasbynens — APPROVED** [review]

**Pull request merged**
